src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -767,25 +768,54 @@ public ISymbolNode DelegateCtor(TypeDesc delegateType, MethodDesc targetMethod,
             TypeAndMethod ctorKey = new TypeAndMethod(delegateType, targetMethod);
             if (!_delegateCtors.TryGetValue(ctorKey, out ctorNode))
             {
-                ModuleToken delegateTypeToken;
-                if (CompilationModuleGroup.ContainsType(delegateType) && delegateType is EcmaType ecmaType)
-                {
-                    delegateTypeToken = new ModuleToken(ecmaType.EcmaModule, (mdToken)MetadataTokens.GetToken(ecmaType.Handle));
-                }
-                else
-                {
-                    // TODO: reverse typedef lookup within the version bubble
-                    throw new NotImplementedException();
-                }
-
                 IMethodNode targetMethodNode = MethodEntrypoint(targetMethod, methodToken, isUnboxingStub: false);
 
                 ctorNode = new DelayLoadHelperImport(this,
                     ILCompiler.DependencyAnalysis.ReadyToRun.ReadyToRunHelper.READYTORUN_HELPER_DelayLoad_Helper,
-                    new DelegateCtorSignature(this, delegateType, delegateTypeToken, targetMethodNode, methodToken));
+                    new DelegateCtorSignature(this, delegateType, default(ModuleToken), targetMethodNode, methodToken));
                 _delegateCtors.Add(ctorKey, ctorNode);
             }
             return ctorNode;
         }
+
+        /// <summary>
+        /// Reverse lookup table mapping external types to reference tokens in the input modules. The table
+        /// gets lazily initialized with all input module references the first time the reverse lookup is
+        /// needed.
+        /// </summary>
+        private Dictionary<TypeDesc, ModuleToken> _typeToRefTokens = null;
+
+        public ModuleToken GetModuleTokenForType(EcmaType type)
```

Comment:
This method is currently only called from SignatureBuilder, where we have `_contextModule`. Should the context module be passed along here so that we get a typeref from the correct "point of view". I think we'll end up generating invalid signatures as-is. For example, if assemblies A and B have a typeref to System.Object and we're building a signature for A (A is the `_contextModule`), we don't want to put a typeref from B in the signature since it'll point at the wrong type most likely.

Reply:
I think we should be able to use a typeref relative to a different module than the one we're compiling right now, we just need to use the 'module override' flag in the signature. I admit I haven't tried this out yet though (in part because the relevant infrastructure isn't yet in place). Based on reading the CoreCLR runtime code I assume we'll somehow need to emit additional fixup cells representing the additional modules within the versioning bubble and encode their indices in the signature.